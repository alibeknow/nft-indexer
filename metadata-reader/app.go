package metadata_reader

import (
	"context"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/zeebo/errs"
	"go.uber.org/zap"
	"golang.org/x/sync/errgroup"

	"metadata-reader/database"
	"metadata-reader/private/httpreader"
	"metadata-reader/queue"
	"metadata-reader/server"
)

var (
	requestPerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "request_per_second_gauge",
			Help: "Number of requests per second",
		},
	)
	messagePerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "message_per_second_gauge",
			Help: "Number of messages per second",
		},
	)
	instantRequestPerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "instant_request_per_second_gauge",
			Help: "Instant number of reqs per second",
		},
	)

	instantMessagePerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "instant_messages_per_second_gauge",
			Help: "Instant number of reqs per second",
		},
	)
)

func init() {
	prometheus.MustRegister(requestPerSecond)
	prometheus.MustRegister(messagePerSecond)
	prometheus.MustRegister(instantRequestPerSecond)
	prometheus.MustRegister(instantMessagePerSecond)
}

type App struct {
	log      *zap.Logger
	db       *database.DB
	srv      *server.Server
	listener *queue.Listener
}

func NewApp(log *zap.Logger, db *database.DB, srv *server.Server, listener *queue.Listener) (*App, error) {
	app := &App{
		log:      log,
		db:       db,
		srv:      srv,
		listener: listener,
	}

	return app, nil
}

func (mr *App) Run(ctx context.Context) error {
	group, ctx := errgroup.WithContext(ctx)

	group.Go(func() error {
		now := time.Now()
		ticker := time.NewTicker(time.Second)
		defer ticker.Stop()

		var (
			reqLast int64
			msgLast int64
		)

		for {
			select {
			case <-ctx.Done():
				return nil
			case <-ticker.C:
				duration := time.Since(now).Seconds()

				reqCount := httpreader.TotalRequestCount()
				requestPerSecond.Set(float64(reqCount) / duration)
				reqDelta := reqCount - reqLast
				reqLast = reqCount
				instantRequestPerSecond.Set(float64(reqDelta))

				msgCount := queue.TotalMessageCount()
				messagePerSecond.Set(float64(msgCount) / duration)
				msgDelta := msgCount - msgLast
				msgLast = msgCount
				instantMessagePerSecond.Set(float64(msgDelta))
			}
		}
	})
	group.Go(func() error {
		return mr.srv.Run(ctx)
	})
	group.Go(func() error {
		return mr.listener.Run(ctx)
	})

	return group.Wait()
}

func (mr *App) Close() error {
	serverErr := mr.srv.Close()
	listenerErr := mr.listener.Close()

	return errs.Combine(serverErr, listenerErr)
}
