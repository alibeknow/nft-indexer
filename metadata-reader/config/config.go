package config

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strings"

	"github.com/spf13/viper"

	"metadata-reader/database"
	"metadata-reader/metadata"
	"metadata-reader/private/logutils"
	"metadata-reader/private/opensearch"
	"metadata-reader/queue"
	"metadata-reader/server"
)

type Config struct {
	Logger   logutils.Config
	Server   server.Config
	Database database.Config
	Queue    queue.Config
	Metadata metadata.Config
	Search   opensearch.Config
}

func LoadConfig(configPaths ...string) (Config, error) {
	v := viper.New()
	v.SetConfigName("config")
	v.SetConfigType("yaml")
	v.SetEnvPrefix("MR")
	v.SetEnvKeyReplacer(strings.NewReplacer(`.`, `_`))
	v.AutomaticEnv()

	v2 := viper.New()
	v2.SetConfigName("config")
	v2.SetConfigType("yaml")
	var buff bytes.Buffer
	m := map[string]interface{}{}
	err := json.NewEncoder(&buff).Encode(Config{})
	if err != nil {
		return Config{}, fmt.Errorf("failed to encode config: %v", err)
	}
	err = json.NewDecoder(&buff).Decode(&m)
	if err != nil {
		return Config{}, fmt.Errorf("failed to decode config: %v", err)
	}
	for key, val := range m {
		v2.Set(key, val)
	}

	for _, path := range configPaths {
		v.AddConfigPath(path)
		v2.AddConfigPath(path)
	}

	if err = v2.SafeWriteConfig(); err != nil {
		if _, ok := err.(viper.ConfigFileAlreadyExistsError); ok {
			log.Println("config.yaml file already exists: skip creating")
		} else {
			return Config{}, fmt.Errorf("failed to write the configuration file: %v", err)
		}
	}

	if err = v.ReadInConfig(); err != nil {
		if _, ok := err.(viper.ConfigFileNotFoundError); ok {
			log.Println("config.yaml ignored")
		} else {
			return Config{}, fmt.Errorf("failed to read the configuration file: %v", err)
		}
	}

	for _, k := range v.AllKeys() {
		value := v.GetString(k)
		v.Set(k, os.ExpandEnv(value))
	}

	var config Config
	err = v.Unmarshal(&config)
	return config, err
}
