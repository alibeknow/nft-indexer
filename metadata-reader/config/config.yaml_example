server:
  address: ":xxx"

database:
  name: "xxx"
  protocol: "mongodb"
  username: "xxx"
  password: "xxx"
  host: "xxx" #should be "localhost" for local use, and "mongodb" for docker
  port: xxx #number
  sslEnable: true/false
  directConnection: true/false

queue:
  protocol: "amqp"
  username: "xxx"
  password: "xxx"
  host: "xxx" #should be "localhost" for local use, and "rabbitmq" for docker
  port: xxx #number
  name: "xxx"

reader:
  workers: xxx #number
  proxyUri: "xxx"
  ipfsGateway: "https://ipfs.io/ipfs/"
  pinataDefaultGateway: "https://gateway.pinata.cloud/ipfs/"

httpreader:
  retries: xxx #number
