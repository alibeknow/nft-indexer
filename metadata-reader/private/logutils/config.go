package logutils

type Config struct {
	Level string
}
