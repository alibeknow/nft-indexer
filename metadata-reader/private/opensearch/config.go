package opensearch

type Config struct {
	Uri      string
	Shards   int
	Replicas int
}
