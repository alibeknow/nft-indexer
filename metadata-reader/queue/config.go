package queue

import "fmt"

type Config struct {
	Protocol string
	Username string
	Password string
	Host     string
	Port     int
	Name     string
	Workers  int
}

func (config *Config) URL() string {
	url := fmt.Sprintf(
		"%s://%s:%s@%s:%d",
		config.Protocol,
		config.Username,
		config.Password,
		config.Host,
		config.Port,
	)

	return url
}
