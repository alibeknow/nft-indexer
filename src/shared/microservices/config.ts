import { IsBoolean, IsEnum, IsInt, IsOptional, IsString } from 'class-validator';

import { CommunicationTransport } from '@shared/microservices';

export enum RabbitProtocol {
  AMQP = 'amqp',
  AMQPS = 'amqps',
}

export interface IMicroservicesConfig {
  transport: CommunicationTransport;
}

export interface ISqsConfig {
  messageDeduplication?: boolean;
  messageGrouping?: boolean;
  queueWorkersCount?: number;
  blocksQueue: string;
  contractsQueue: string;
  metadataQueue: string;
}

export type QueueType = keyof Pick<ISqsConfig, 'blocksQueue' | 'contractsQueue' | 'metadataQueue'>;

export interface IRabbitConfig {
  protocol: RabbitProtocol;
  host: string;
  port: number;
  username?: string;
  password?: string;
  blocksQueue: string;
  contractsQueue: string;
  metadataQueue: string;
}

export class MicroservicesConfigValidator implements IMicroservicesConfig {
  @IsEnum(CommunicationTransport)
  readonly transport!: CommunicationTransport;
}

export class SqsConfigValidator implements ISqsConfig {
  @IsBoolean()
  @IsOptional()
  readonly messageDeduplication?: boolean;

  @IsBoolean()
  @IsOptional()
  readonly messageGrouping?: boolean;

  @IsInt()
  @IsOptional()
  readonly queueWorkersCount?: number;

  @IsString()
  readonly blocksQueue!: string;

  @IsString()
  readonly contractsQueue!: string;

  @IsString()
  readonly metadataQueue!: string;
}

export class RabbitConfigValidator implements IRabbitConfig {
  @IsEnum(RabbitProtocol)
  readonly protocol!: RabbitProtocol;

  @IsString()
  readonly host!: string;

  @IsInt()
  readonly port!: number;

  @IsString()
  @IsOptional()
  readonly username?: string;

  @IsString()
  @IsOptional()
  readonly password?: string;

  @IsString()
  readonly blocksQueue!: string;

  @IsString()
  readonly contractsQueue!: string;

  @IsString()
  readonly metadataQueue!: string;
}

export const getMicroservicesConfig = (): IMicroservicesConfig => ({
  transport: process.env.COMMUNICATION_TRANSPORT as CommunicationTransport,
});

export const getSqsConfig = (): ISqsConfig => ({
  messageDeduplication: process.env.SQS_MESSAGE_DEDUPLICATION === 'true',
  messageGrouping: process.env.SQS_MESSAGE_GROUPING === 'true',
  queueWorkersCount: parseInt(`${process.env.SQS_QUEUE_WORKERS_COUNT ?? 1}`, 10) as number,
  blocksQueue: process.env.SQS_BLOCKS_QUEUE as string,
  contractsQueue: process.env.SQS_CONTRACTS_QUEUE as string,
  metadataQueue: process.env.SQS_METADATA_QUEUE as string,
});

export const getRabbitConfig = (): IRabbitConfig => ({
  protocol: process.env.RABBIT_PROTOCOL as RabbitProtocol,
  host: process.env.RABBIT_HOST as string,
  port: parseInt(`${process.env.RABBIT_PORT}`, 10) as number,
  username: process.env.RABBIT_USERNAME as string,
  password: process.env.RABBIT_PASSWORD as string,
  blocksQueue: process.env.RABBIT_BLOCK_READ_QUEUE as string,
  contractsQueue: process.env.RABBIT_CONTRACT_READ_QUEUE as string,
  metadataQueue: process.env.RABBIT_METADATA_READ_QUEUE as string,
});

export const getRabbitUrl = (config: IRabbitConfig): string => {
  const { protocol, username, password, host, port } = config;

  if (username && password) {
    return `${protocol}://${encodeURIComponent(username)}:${encodeURIComponent(password)}@${host}:${port}`;
  }

  return `${protocol}://${host}:${port}`;
};
