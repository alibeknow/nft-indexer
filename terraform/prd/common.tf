data "terraform_remote_state" "shared_application" {
  backend = "s3"
  config = {
    bucket  = "ledger-${local.env_mapping[var.env]}-tfstates"
    key     = local.tfs_mapping[var.env].shared
    region  = var.region
    profile = var.profile
  }
}

data "terraform_remote_state" "global" {
  backend = "s3"
  config = {
    bucket  = "ledger-${local.env_mapping[var.env]}-tfstates"
    key     = local.tfs_mapping[var.env].global
    region  = var.region
    profile = var.profile
  }
}

data "terraform_remote_state" "platform" {
  count   = var.env == "stg" ? 0 : 1
  backend = "s3"
  config = {
    bucket  = "ledger-${local.env_mapping[var.env]}-tfstates"
    key     = "2020-platform/platform.tfstate"
    region  = var.region
    profile = var.profile
  }
}
