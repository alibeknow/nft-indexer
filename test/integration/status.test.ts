process.env.APP_VERSION = '0.0.1';

import { HttpServer, INestApplication } from '@nestjs/common';
import { Test } from '@nestjs/testing';
import { ConfigModule } from '@nestjs/config';
import request from 'supertest';
import { LoggerModule } from 'nestjs-pino';
import { getBaseConfig } from '@shared/baseconfig';
import { getNestApp } from '../helpers/app';
import { StatusController } from '@shared/status';

jest.setTimeout(50000);

describe('[Status endpoint]', () => {
  let app: INestApplication;
  let httpServer: HttpServer;

  beforeAll(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [
        LoggerModule.forRoot(),
        ConfigModule.forRoot({
          load: [ getBaseConfig ],
          isGlobal: true,
        }),
      ],
      controllers: [ StatusController ],
    }).compile();

    app = await getNestApp(moduleRef);
    httpServer = app.getHttpServer();
  });

  describe('[GET /_health]', () => {
    it('should return OK and status code 200', async () => {
      const actualResult = await request(httpServer).get('/_health').set('Accept', 'application/json');

      expect(actualResult.statusCode).toBe(200);
      expect(actualResult.body).toMatchSnapshot();
    });
  });

  describe('[GET /_version]', () => {
    it('should return version and status code 200', async () => {
      const actualResult = await request(httpServer).get('/_version').set('Accept', 'application/json');

      expect(actualResult.statusCode).toBe(200);
      expect(actualResult.body).toMatchSnapshot();
    });
  });
});
