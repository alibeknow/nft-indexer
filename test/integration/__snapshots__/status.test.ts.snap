// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Status endpoint] [GET /_health] should return OK and status code 200 1`] = `
{
  "result": "OK",
  "statusCode": 200,
}
`;

exports[`[Status endpoint] [GET /_version] should return version and status code 200 1`] = `
{
  "result": "0.0.1",
  "statusCode": 200,
}
`;
